www/js/diary/LabelTab.tsx

```diff
@@ -0,0 +1,374 @@
+      readUnprocessedPromise = Timeline.readUnprocessedTrips(pipelineRange.end_ts, nowTs, allTrips);
+    } else {
+      readUnprocessedPromise = Promise.resolve([]);
+    }
+    return await Promise.all([readCompositePromise, readUnprocessedPromise]);
+  };
+
+  function fillPlacesForTripAsync(trip) {
+    const fillPromises = [
+      placeLimiter.schedule(() =>
+        DiaryHelper.getNominatimLocName(trip.start_loc)),
+      placeLimiter.schedule(() =>
+        DiaryHelper.getNominatimLocName(trip.end_loc)),
+    ];
+    Promise.all(fillPromises).then(function ([startName, endName]) {
+      if (trip.start_confirmed_place) {
+        trip.start_confirmed_place.display_name = startName;
+        trip.start_confirmed_place.onChanged?.(); /* Temporary hack, see below */
+      }
+      trip.start_display_name = startName;
+      trip.end_display_name = endName;
+      if (trip.end_confirmed_place) {
+        trip.end_confirmed_place.display_name = endName;
+        trip.end_confirmed_place.onChanged?.(); /* Temporary hack, see below */
+      }
+      trip.onChanged?.(); /* Temporary hack for React to update when
+                                      data changes in Angular.
+                                    Will not be needed later. */
+    });
+  }
+
+  useEffect(() => {
+    if (!displayTrips?.length) return;
+    const newListEntries = [];
+    displayTrips.forEach((cTrip) => {
+      const start_place = cTrip.start_confirmed_place;
+      const end_place = cTrip.end_confirmed_place;
+
+      // Add start place to the list, if not already present
+      let isInList = newListEntries.find(e => e._id?.$oid == start_place?._id.$oid);
+      if (showPlaces && start_place && !isInList) {
+        // Only display places with duration >= 60 seconds, or with no duration (i.e. currently ongoing)
+        if (isNaN(start_place.duration) || start_place.duration >= 60) {
+          newListEntries.push(start_place);
+        }
+      }
+
+      /* don't display untracked time if the trips that came before and
+          after it are not displayed */
+      if (cTrip.key.includes('untracked')) {
+        const prevTrip = allTrips[allTrips.indexOf(cTrip) - 1];
+        const nextTrip = allTrips[allTrips.indexOf(cTrip) + 1];
+        const prevTripDisplayed = displayTrips.includes(prevTrip);
+        const nextTripDisplayed = displayTrips.includes(nextTrip);
+        if (prevTrip && !prevTripDisplayed || nextTrip && !nextTripDisplayed) {
+          return;
+        }
+      }
+
+      // Add trip to the list
+      newListEntries.push(cTrip);
+
+      // Add end place to the list
+      if (showPlaces && end_place) {
+        // Only display places with duration >= 60 seconds, or with no duration (i.e. currently ongoing)
+        if (isNaN(end_place.duration) || end_place.duration >= 60) {
+          newListEntries.push(end_place);
+        }
+      }
+    });
+    setListEntries(newListEntries);
+  }, [displayTrips]);
+
+  useEffect(() => {
```

Comment:
Definitely something to consider.
I would guess that roughly half of this file is just for retrieving, filtering, and formatting the trips.
And that's code that shouldn't really be in LabelTab; it can live elsewhere

Reply:
I refactored this flow in e8090a6ddc1c99eba8e570090802c13997896063. I did not explicitly structure it as a pipeline, but I do think it is simpler and more comprehensible now